zerver/models.py

```diff
@@ -183,6 +183,48 @@ def query_for_ids(
     return query
 
 
+def get_column_values_from_single_table_using_id_lookup(
+    *, columns: List[str], table: str, id_field: str, ids: List[int], order_by_columns: List[str]
+) -> List[Dict[str, object]]:
+    """
+    Use this function somewhat judiciously, because it does impose a little bit
+    of burden on security auditors to make sure our callers are passing
+    `columns` that are not from user input and therefore not vulnerable to
+    injection attacks.
+
+    But this function is safe and fast. The Django ORM takes almost half a
+    millisecond to produce really simple queries that just fetch a bunch of
+    columns from a single table based on a bunch of ids. This function is
+    better.
+    """
+    if len(ids) == 0:
+        return []
+
+    column_list = ", ".join(columns)
+
+    assert table.startswith("zerver_")
+
+    for id in ids:
+        assert type(id) == int
+
+    id_list = ", ".join(str(id) for id in ids)
+
+    cursor = connection.cursor()
+
+    # ruff: noqa: S608
+    sql = f"SELECT {column_list} FROM {table} WHERE {id_field} in ({id_list})"
```

Comment:
From a security perspective, I can’t endorse any code that plays fast and loose with string interpolation to generate SQL. No matter how much you implore the caller in the docstring to uphold the extra safety preconditions, someone’s going to use this without carefully reading the docstring, or copy code to a different context where the safety preconditions don’t hold.

Use the [`psycopg2.sql` API](https://www.psycopg.org/docs/sql.html) to build SQL queries safely without preconditions, like we do in various places.

Reply:
Ok, I probably won't get to this for a couple days, but I can appreciate the caution here.  I am familiar with the API from  `query_for_ids`, so I can fix this up the next time I'm working, and I just want to double check the benchmarks.